### ISSUE TITLE
Improve interaction with Pandas: remove '#' from header of StateDataReporter?

### ISSUE THREAD

**Issue by tovrstra** [openmm/openmm#2403]
When reading a CSV file generated by the `StateDataReporter` with Pandas (or spreadsheet software) the header is not parsed correctly. More specifically, the header of the first column contains spurious characters. For example, try to following to load the output of the `StateDataReporter`:

```python
import pandas
df = pandas.read_csv("openmm.csv")
print(df.head())
```

This shows the problem with the header:

```
   #"Step"  Potential Energy (kJ/mole)  Temperature (K)
0        1                    1.888971       238.107266
1        2                    4.637080       139.250504
2        3                    4.249358       150.950951
3        4                    2.043082       228.325732
4        5                    1.896116       235.380171
```

This is essentially caused by a redundant comment character, `#`, written out by `StateDataReporter` in front of the header line.

There are no options in Pandas to ignore just this single comment character, such that the header is read correctly. (See https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html) I guess it is not a common practice (in data science at least) to comment out the header of a CSV file.

Would it be ok to just remove the `#` character from the following line of code?

https://github.com/openmm/openmm/blob/f9106ddb6a0d72c960c73b23ed0e07cd7190a80f/wrappers/python/simtk/openmm/app/statedatareporter.py#L184

I'm not sure if this would break compatibility with other post-processing tools? Alternatively, the comment character could be turned into an optional argument, allowing one to specify `comment=""`.

This is related to issue #325.

**Comment by peastman** [comment]
Including that character was kind of a compromise.  When I first implemented the reporter I tried loading the output into various programs, and that gave the least overall broken behavior.  It's been a long time, so I don't remember the details.  I'm pretty sure Matlab completely failed to read the file if I didn't include it.

**Comment by tovrstra** [comment]
I understand and I can imagine not all programs can handle a header line and need to skip it instead.

I checked Matlabs `csvread` (https://nl.mathworks.com/help/matlab/ref/csvread.html), which says indeed `The file must contain only numeric values`. (This function is no longer recommended, but that doesn't matter.) I recall now that some Gromacs tools have similar limitations. So ditching the comment character is not going to work for everyone.

Would an extra optional argument to `StateDataReporter` still be welcome? If yes, should related issues like #325 be tackled too?

**Comment by peastman** [comment]
An extra option would be a reasonable solution.  What about a `format` option that accepts values like `'matlab'`, `'pandas'`, `'fixed'`, etc.?

**Comment by tovrstra** [comment]
A simple `format` option would indeed be very practical.

For the implementation, I tried to come with a proposal that can be easily extended later with more formats or more formatting options, yet without adding a large number of arguments. I would suggest something along the following lines.

- Add the `format` option to the constructor of `StateDataReporter`, with `'matlab'` as default. In addition to the strings, `'matlab'`, `'pandas'`, `'fixed'`, ... it can also take a dictionary with detailed settings. The strings are then just shorthands for built-in dictionaries. Several things will be affected by the format option: presence of a comment in front of the header, separator, fixed (or not) column width, formatting of the numbers and a flag to put units on second line. This should be enough to support the style of `ProgressReporter` in #325. The dictionary idea may seem overkill at first, but without them, the code would contain a large number of `if format == "something"` conditionals, which would not be very maintainable.

- Change the default of `separator` to `None`, meaning that the separator is determined by the `format` argument. When this argument is provided, it overrides the separator implied by the `format` option.

When no `format` argument is given, the new code should reproduce the behavior of the current implementation.

This is what the dictionaries can look like:

```python

FORMAT_MATLAB = {
    'separator': ',',
    'comment': '#',
    'fixedColumnWidth': False,
    'wrapUnits': False,  # put units in the first header line
}

FORMAT_PANDAS = {
    'separator': ',',
    'comment': '',
    'fixedColumnWidth': False,
    'wrapUnits': False,  # put units in the first header line
}

def formatTiming(seconds):
    days = seconds//86400
    seconds -= days*86400
    hours = seconds//3600
    seconds -= hours*3600
    minutes = seconds//60
    seconds -= minutes*60
    if days > 0:
        return "%d:%d:%02d:%02d" % (days, hours, minutes, seconds)
    elif hours > 0:
        return "%d:%02d:%02d" % (hours, minutes, seconds)
    elif minutes > 0:
        return "%d:%02d" % (minutes, seconds)
    else:
        return "0:%02d" % seconds

def formatProgress(progress):
    return "{:6.2f}%".format(progress*100)

FORMAT_FIXED = {
    'separator': ' ',
    'comment': '',
    'fixedColumnWidth': True,
    'wrapUnits': True,  # put units on second line
    # All keys starting with 'format' should map to a function,
    # which takes an int or float and returns a string.
    # When a formatXxx is missing, it can default to `repr`.
    'formatStep': "{:13d}".format,
    'formatEnergy': "{:15.2f}".format,
    'formatTime': "{:15.3f}".format,
    'formatTemperature': "{:12.1f}".format,
    'formatVolume': "{:15.4f}".format,
    'formatDensity': "{:9.4f}".format,
    'formatTiming': formatTiming,
    'formatProgress': formatProgress,
    'formatSpeed': "{:8.2f}".format,
}
```

The actual widths of the columns, in case of fixed widths, can be determined upfront from the header strings and by calling the formatters with argument `0`.

This approach also guarantees that in case of `'matlab'` (and `'pandas'`), the numerical data are indeed just numbers, which MATLAB's `csvread` can handle. In general, this also facilitates plotting.

**Comment by peastman** [comment]
Sorry for not replying sooner.  I wanted to take some time to think about this proposal.  After debating it a bit, I think it's both more general than we need and at the same time not general enough.  Users don't really need that level of fine control over the output formatting.  For nearly all purposes, just selecting between a few options should be fine.  On the other hand, this design would firmly tie us to the idea that output formats are always delimited text.  What if someone wants to add `format='netcdf'`?  I don't know if that's likely to happen, but it's the sort of thing someone might reasonably ask for.

**Comment by tovrstra** [comment]
Apologies for dropping the ball. I also prefer thinking before commenting (and there were few other things I had to take care of).

I agree that binary formats for the `StateDataReporter` could be of interest, but I would not implemented them literally inside the `StateDataReporter` class. `NetCDFReporter` and `HDF5Reporter` implementations can be found in [parmed](https://parmed.github.io/ParmEd/html/api/parmed/parmed.openmm.reporters.html) and [MDTraj](http://mdtraj.org/1.9.3/api/reporters.html). Those from MDTraj are nearly sufficient as binary equivalents of the `StateDataReporter`. If one wants to support, for example, a `format='netcdf'` option in `StateDataReporter`, I'd suggest to handle this option in the `__new__` method and return an equivalent `NetCDFReporter` instance. This way the implementation of each reporter remains tied to a specific format (which is easier to manage from a programming perspective) while the user still has to convenience of a powerful `format` option in `StateDataReporter`.

**Comment by peastman** [comment]
For the moment, lets keep `format` as just as string.  If we decide we want more flexibility in the future it's easy to add it later.

**Comment by tovrstra** [comment]
Agreed. It is not essential to offer such flexibility to the user. By keeping the API simple, there is more flexibility in the future to make changes.

I'll experiment a little to see if it makes sense to use the dictionaries under to hood to obtain a simple implementation.
